pydropsonde/processor.py

```diff
@@ -2098,18 +2169,66 @@ def get_circle_times_from_segmentation(self, yaml_file: str = None):
         segmentation = rr.get_flight_segmentation(yaml_file)
         platform_ids = set(self.l3_ds.platform_id.values)
         flight_ids = set(self.l3_ds.flight_id.values)
-        self.segments = [
+        self.segments = sorted(
+            [
+                {
+                    **s,
+                    "platform_id": platform_id,
+                    "flight_id": flight_id,
+                }
+                for platform_id in platform_ids
+                for flight_id in flight_ids
+                for s in segmentation.get(platform_id, {})
+                .get(flight_id, {})
+                .get("segments", [])
+                if "circle" in s["kinds"]
+            ],
+            key=lambda s: s["start"],
+        )
+
+        return self
+
+    def get_l4_dir(self, l4_dir: str = None):
+        if l4_dir:
+            self.l4_dir = l4_dir
+        elif self.circles is not None:
+            self.l4_dir = self.l3_dir.replace("Level_3", "Level_4")
+        else:
+            raise ValueError("No circles and no l4 directory given, cannot continue")
+        return self
+
+    def get_l4_filename(self, l4_filename: str = None):
+        if l4_filename is None:
+            l4_filename = hh.l4_filename
+        else:
+            l4_filename = l4_filename
+
+        self.l4_filename = l4_filename
+        return self
+
+    def write_l4(self, l4_dir: str = None, _interim_l4_ds: xr.Dataset = None):
+        if l4_dir is None:
+            l4_dir = self.l4_dir
+        ds = self._interim_l4_ds
+        history = getattr(self, "history", "")
+        history = (
+            history
+            + datetime.now(timezone.utc).isoformat()
+            + f" level4 concatenation with pydropsonde {__version__} \n"
+        )
+        object.__setattr__(self, "history", history)
+        ds.attrs.update(
             {
-                **s,
-                "platform_id": platform_id,
-                "flight_id": flight_id,
+                "history": history,
+                "title": ds.attrs.get("title", "Dropsonde Data") + " Level 4",
```

Comment:
The `ds` within this function seems to inherit the title from the Level3 dataset such that you only need to exchange the 3 by the 4 at the end. Otherwise you will end up with the default of `Dropsonde Data Level 3 Level 4`. Could be a valid idea, too, as we have level 3 data in there as well :) but I'd suggest to replace the 3 by 4.
```suggestion
                "title": ds.attrs.get("title", "Dropsonde Data")[:-1] + "4",
```

Reply:
Yes! That's more robust and way better :)

Reply:
Update: a more general discussion implemented in PR #176 covers the present thread and therefore we can close this issue here